ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -254,68 +260,84 @@ getBlockImpl env@VolatileDBEnv{..} slot = do
--- collection of files.
+-- If there is an error after (1) or after (2) we should make sure that when
+-- we reopen a db from scratch, it can succesfully recover if it does not
+-- find an empty file to write and all other files are full.
+-- We should also make sure that the fs can be recovered if we get an
+-- exception/error at any moment and that we are left with an empty Internal
+-- State.
+-- We should be careful about not leaking open fds when we open a new file,
+-- since this can affect garbage collection of files.
 putBlockImpl :: forall m blockId. (IOLike m, Ord blockId)
              => VolatileDBEnv m blockId
              -> BlockInfo blockId
              -> BS.Builder
              -> m ()
-putBlockImpl env@VolatileDBEnv{..} BlockInfo{..} builder = do
-    modifyState env $ \hasFS@HasFS{..} st@InternalState{..} -> do
-        case Map.lookup bbid _currentRevMap of
-            Just _ -> return (st, ()) -- trying to put an existing block is a no-op.
-            Nothing -> do
-                let fileInfo = fromMaybe
-                        (error "Volatile db invariant violation: Current write file not found in Index.")
-                        (Map.lookup _currentWritePath _currentMap)
-                bytesWritten <- hPut hasFS _currentWriteHandle builder
-                let
-                    fileInfo' = FileInfo.addSlot bslot _currentWriteOffset (FileSlotInfo (fromIntegral bytesWritten) bbid) fileInfo
-                    mp = Map.insert _currentWritePath fileInfo' _currentMap
-                    internalBlockInfo' = InternalBlockInfo {
-                          ibFile       = _currentWritePath
-                        , ibSlotOffset = _currentWriteOffset
-                        , ibBlockSize  = bytesWritten
-                        , ibSlot       = bslot
-                        , ibPreBid     = bpreBid
-                    }
-                    revMp = Map.insert bbid internalBlockInfo' _currentRevMap
-                    st' = st {
-                          _currentWriteOffset = _currentWriteOffset + fromIntegral bytesWritten
-                        , _currentMap         = mp
-                        , _currentRevMap      = revMp
-                        , _currentSuccMap     = insertMapSet _currentSuccMap (bbid, bpreBid)
-                        , _currentMaxSlotNo   = _currentMaxSlotNo `max` MaxSlotNo bslot
-                    }
-                if not (FileInfo.isFull _maxBlocksPerFile fileInfo')
-                then return (st', ())
-                else (\s -> (s,())) <$> nextFile hasFS _dbErr env st'
+putBlockImpl env@VolatileDBEnv{..} binfo@BlockInfo{..} builder =
+  modifyState env $ \hasFS@HasFS{..} st@InternalState{..} ->
+    if Map.member bbid _currentRevMap
+    then return (st, ()) -- putting an existing block is a no-op.
+    else do
+        bytesWritten <- hPut hasFS _currentWriteHandle builder
+        putBlockImpl' env binfo hasFS st bytesWritten
```

Comment:
@mrBliss I am a bit confused about 4 vs 2 spaces. in which cases, do we have to use 4 but not 2?

Reply:
The guideline in this repository is: use 2 spaces for indentation, unless there's a `where`-clause for the given block, in which case you should indent the `where` keyword 2 spaces and the block above the `where` keyword and the bindings below it with 4 spaces.

So:
```haskell
foo ... =
  bla di bla
```
but
```haskell
foo ... =
    bla di bla
  where
    bla = ..
```
In nested code blocks, the same principle holds.

I intentionally deviate from this guideline in one case: I indent the body of a top-level function (or a big local function) with 4 spaces by default, even without a `where`-clause, so that in case we add one, we don't have to reindent.